docs/proposals/metadata.md

```diff
@@ -0,0 +1,152 @@
+# Image Metadata
+
+## Abstract
+
+A proposal for the additional metadata for Docker images which will provide more
+context about the content of the image and define relations between Docker
+images.
+
+## Motivation
+
+Image metadata are needed to do automatic generation of the OpenShift v3
+resource templates, like PodTemplates, ReplicationControllers, BuildConfigs
+etc.  Automatic generation of a PodTemplate is also required for
+Deployments.
+
+## Constraints and Assumptions
+
+For the time being, there is no good way to define metadata for Docker
+images other than specifying extra environment variables. There is upstream
+work currently being done to fix this situation and we should switch to that in
+the implementation in the future.
+
+This document only defines the metadata needed by the current set of use cases.
+Additional metadata and/or use cases may be added in the future.
+
+## Use cases
+
+1. As an author of the Docker image, I want to specify a flag that will tell
+   whether the image I build can be used as an STI builder image, so that when
+   somebody uses my image in OpenShift v3, the BuildConfig generated for this image
+   will default the build strategy to STI. For this, I only need to specify if my
+   entrypoint accepts a 'tar' stream that contains the application source as an input.
+
+2. As an author of the Docker image, I want to specify list of services that the
+   image I'm building will provide, so in OpenShift v3 we can automatically
+   setup Routes, Services and health-checks for this image. For example, in case
+   I'm building 'mysql' image, I can specify `PUBLISHES=mysql/tcp/3128` so when
+   the image is deployed, we automatically setup a Service for it, add health-check
+   for 'mysql' service type and wire the Route for the 3128 port.
+
+3. As an user of OpenShift v3 I want to be able to say what services my
+   application will consume, so the platform can find the corresponding Docker
+   images and make the services available for my application.
+   For example, as an user I can say I want to have 'mysql' service available in
+   my application, so the platform will check the ImageRepositories for the
+   Image that have `PUBLISHES="mysql/tcp/3128"` and add it to my deployment.
+
+4. As an author of the Docker image, I want to specify what _secrets_ the image
```

Comment:
That mechanism is templates.

----- Original Message -----
> > +
> > +2. As an author of the Docker image, I want to specify list of services
> > that the
> > +   image I'm building will provide, so in OpenShift v3 we can
> > automatically
> > +   setup Routes, Services and health-checks for this image. For example,
> > in case
> > +   I'm building 'mysql' image, I can specify `PUBLISHES=mysql/tcp/3128` so
> > when
> > +   the image is deployed, we automatically setup a Service for it, add
> > health-check
> > +   for 'mysql' service type and wire the Route for the 3128 port.
> > +
> > +3. As an user of OpenShift v3 I want to be able to say what services my
> > +   application will consume, so the platform can find the corresponding
> > Docker
> > +   images and make the services available for my application.
> > +   For example, as an user I can say I want to have 'mysql' service
> > available in
> > +   my application, so the platform will check the ImageRepositories for
> > the
> > +   Image that have `PUBLISHES="mysql/tcp/3128"` and add it to my
> > deployment.
> > +
> > +4. As an author of the Docker image, I want to specify what _secrets_ the
> > image
> 
> this is going to be the mechanism for getting DB passwords and such, is it
> not?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/906/files#r24196840